From: Maximilian Weigand <mweigand@mweigand.net>
Date: Tue, 6 Jun 2023 10:43:05 +0200
Subject: drm/rockchip: ebc: add additional panel modes for the eink_ed103tc2
 panel

In total, define three panel modes for the panel with different dclk
clock, resulting in 63.74, 70.12, and 85 Hz panel refresh rates.

WARNING: The other panel timing parameters, especially hskew, have been
altered and are most probably WRONG. This still needs to be tweaked!
---
 drivers/gpu/drm/panel/panel-simple.c | 78 ++++++++++++++++++++++++++----------
 1 file changed, 56 insertions(+), 22 deletions(-)

diff --git a/drivers/gpu/drm/panel/panel-simple.c b/drivers/gpu/drm/panel/panel-simple.c
index f86b5832..0f6b2f3 100644
--- a/drivers/gpu/drm/panel/panel-simple.c
+++ b/drivers/gpu/drm/panel/panel-simple.c
@@ -2095,31 +2095,65 @@ static const struct panel_desc edt_etmv570g2dhu = {
 	.connector_type = DRM_MODE_CONNECTOR_DPI,
 };
 
-static const struct drm_display_mode eink_ed103tc2_mode = {
-	/* this will lead to dclk == 250 Mhz */
-	/* .clock = 266693, */
-	/* request dclk == 200 MHz */
-	.clock = 200000,
-	.hdisplay = 1872,
-	/* hsync_start, for linux timing order, should be display + fp
-	 * according to the data sheet, display = ldl == 234 [dclk] = 1872 pixels, and fp == lel == 7 [dclk] == 56 pixels
-	 */
-	.hsync_start = 1872 + 56,
-	/* data sheet days sync == lsl == 18 sclk = 18 * 8 pixels = 144 pixels */
-	.hsync_end = 1872 + 56 + 144,
-	.htotal = 1872 + 56 + 144 + 136,
-	.hskew = 136,
-	.vdisplay = 1404,
-	.vsync_start = 1404 + 12,
-	.vsync_end = 1404 + 12 + 1,
-	.vtotal = 1404 + 12 + 1 + 4,
-	.flags = DRM_MODE_FLAG_PHSYNC | DRM_MODE_FLAG_NVSYNC |
-		 DRM_MODE_FLAG_HSKEW | DRM_MODE_FLAG_CLKDIV2,
+static const struct drm_display_mode eink_ed103tc2_mode[] ={
+	{
+		/* this will lead to dclk == 250 Mhz */
+		/* .clock = 266693, */
+		/* request dclk == 200 MHz */
+		.clock = 200000,
+		.hdisplay = 1872,
+		/* hsync_start, for linux timing order, should be display + fp
+		 * according to the data sheet, display = ldl == 234 [dclk] = 1872 pixels, and fp == lel == 7 [dclk] == 56 pixels
+		 */
+		/* data sheet days sync == lsl == 18 sclk = 18 * 8 pixels = 144 pixels */
+		.hsync_start = 1872 + 56,
+		.hsync_end = 1872 + 56 + 144,
+		.htotal = 1872 + 56 + 144 + 136,
+		/* .hsync_start = 1872 + 184, */
+		/* .hsync_end = 1872 + 184 + 88, */
+		/* .htotal = 1872 + 184 + 88 + 64, */
+		/* .hskew = 136, */
+		.hskew = 64,
+		.vdisplay = 1404,
+		.vsync_start = 1404 + 12,
+		.vsync_end = 1404 + 12 + 1,
+		.vtotal = 1404 + 12 + 1 + 4,
+		.flags = DRM_MODE_FLAG_PHSYNC | DRM_MODE_FLAG_NVSYNC |
+			 DRM_MODE_FLAG_HSKEW | DRM_MODE_FLAG_CLKDIV2,
+	},
+	{
+		.clock = 220000,
+		.hdisplay = 1872,
+		.hsync_start = 1872 + 56,
+		.hsync_end = 1872 + 56 + 144,
+		.htotal = 1872 + 56 + 144 + 136,
+		.hskew = 64,
+		.vdisplay = 1404,
+		.vsync_start = 1404 + 12,
+		.vsync_end = 1404 + 12 + 1,
+		.vtotal = 1404 + 12 + 1 + 4,
+		.flags = DRM_MODE_FLAG_PHSYNC | DRM_MODE_FLAG_NVSYNC |
+			 DRM_MODE_FLAG_HSKEW | DRM_MODE_FLAG_CLKDIV2,
+	},
+	{
+		.clock = 266693,
+		.hdisplay = 1872,
+		.hsync_start = 1872 + 56,
+		.hsync_end = 1872 + 56 + 144,
+		.htotal = 1872 + 56 + 144 + 136,
+		.hskew = 64,
+		.vdisplay = 1404,
+		.vsync_start = 1404 + 12,
+		.vsync_end = 1404 + 12 + 1,
+		.vtotal = 1404 + 12 + 1 + 4,
+		.flags = DRM_MODE_FLAG_PHSYNC | DRM_MODE_FLAG_NVSYNC |
+			 DRM_MODE_FLAG_HSKEW | DRM_MODE_FLAG_CLKDIV2,
+	},
 };
 
 static const struct panel_desc eink_ed103tc2 = {
-	.modes = &eink_ed103tc2_mode,
-	.num_modes = 1,
+	.modes = eink_ed103tc2_mode,
+	.num_modes = 3,
 	.bpc = 4,
 	.size = {
 		.width = 210,
